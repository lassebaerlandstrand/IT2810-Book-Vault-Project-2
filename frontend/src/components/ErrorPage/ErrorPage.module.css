.container {
  position: relative;
  padding-top: rem(60px);
  padding-bottom: rem(60px);
}

.image {
  position: absolute;
  opacity: 0.8;
  color: light-dark(var(--mantine-color-gray-2), var(--mantine-color-dark-6));
  width: 80%;
  top: 40%;
  left: 50%;
  transform: translate(-50%, -50%);
}

.content {
  padding-top: rem(220px);
  position: relative;
  z-index: 1;

  @media (max-width: $mantine-breakpoint-sm) {
    padding-top: rem(120px);
  }
}

.title {
  text-align: center;
  font-weight: 900;
  font-size: rem(38px);

  @media (max-width: $mantine-breakpoint-sm) {
    font-size: rem(32px);
  }
}

.description {
  max-width: rem(540px);
  margin: auto;
  margin-top: var(--mantine-spacing-xl);
  margin-bottom: calc(var(--mantine-spacing-xl) * 1.5);
}
