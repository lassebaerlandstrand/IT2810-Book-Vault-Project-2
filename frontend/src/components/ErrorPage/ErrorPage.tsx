import { isRouteErrorResponse, Link, useRouteError } from 'react-router-dom';
import {
  Button,
  Center,
  Code,
  Collapse,
  Container,
  Group,
  Stack,
  Text,
  Title,
} from '@mantine/core';
import { useDisclosure } from '@mantine/hooks';
import { Illustration404 } from './Illustration404';
import styles from './ErrorPage.module.css';

type ErrorPageProps = {
  title?: string;
  description?: string;
  link?: string;
};

export function ErrorPage({ title, description, link }: ErrorPageProps) {
  const error = useRouteError();

  if (isRouteErrorResponse(error) && error.status === 404) {
    return <Error404 title={title} description={description} link={link} />;
  }

  return <ErrorUnexpectedError error={error as Error} />;
}

const ErrorUnexpectedError = ({ error }: { error: Error }) => {
  const [opened, { toggle }] = useDisclosure(true);

  const errorMessage = error?.message || 'Unknown error';
  const errorStack = error?.stack || 'No stack trace available';

  return (
    <Container className={styles.container}>
      <Container className={styles.content}>
        <Title className={styles.title}>Unexpected Application Error!</Title>
        <Text c="dimmed" size="lg" ta="center" className={styles.description}>
          Something went wrong, please try again.
        </Text>
        <Center>
          <Link to={'/'}>
            <Button size="md">Take me back to safe ground</Button>
          </Link>
        </Center>
        <Center mt="xl">
          <Stack align="center">
            <Button onClick={toggle} w="fit-content">
              {opened ? 'Hide' : 'Open'} error message
            </Button>
            <Collapse in={opened}>
              <Code block mt="sm">
                {errorMessage}
              </Code>
              <Code block mt="sm">
                {errorStack}
              </Code>
            </Collapse>
          </Stack>
        </Center>
      </Container>
    </Container>
  );
};

const Error404 = ({
  title = 'Nothing to see here',
  description = 'Page you are trying to open does not exist. You may have mistyped the address, or the page has been moved to another URL.',
  link = '/',
}: ErrorPageProps) => {
  return (
    <Container className={styles.container}>
      <Illustration404 className={styles.image} />
      <Container className={styles.content}>
        <Title className={styles.title}>{title}</Title>
        <Text c="dimmed" size="lg" ta="center" className={styles.description}>
          {description}
        </Text>
        <Group justify="center">
          <Link to={link}>
            <Button size="md">Take me back to safe ground</Button>
          </Link>
        </Group>
      </Container>
    </Container>
  );
};
