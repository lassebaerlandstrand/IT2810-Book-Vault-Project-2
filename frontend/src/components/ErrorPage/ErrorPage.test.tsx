// ErrorPage.test.tsx
import { fireEvent, render, screen } from '@test-utils';
import { isRouteErrorResponse, MemoryRouter, useRouteError } from 'react-router-dom';
import { describe, expect, it, vi } from 'vitest';
import { Error404, ErrorPage } from './ErrorPage';

vi.mock('react-router-dom', async () => {
  const actual = await vi.importActual('react-router-dom');
  return {
    ...actual,
    useRouteError: vi.fn(),
    isRouteErrorResponse: vi.fn(),
    Link: ({ children }: any) => <a href="/">{children}</a>, // Mock <Link> for simplicity
  };
});

describe('ErrorPage Component', () => {
  it('renders 404 error page when error status is 404', () => {
    (useRouteError as jest.Mock).mockReturnValue({
      status: 404,
    });
    (isRouteErrorResponse as unknown as jest.Mock).mockReturnValue(true);

    const container = render(
      <MemoryRouter>
        <ErrorPage />
      </MemoryRouter>
    );

    expect(container.baseElement.getElementsByTagName('svg')[0]).toBeInTheDocument();
    expect(screen.getByText('Nothing to see here')).toBeInTheDocument();
    expect(screen.getByText(/page you are trying to open does not exist/i)).toBeInTheDocument();
    expect(screen.getByRole('link', { name: /take me back to safe ground/i })).toBeInTheDocument();
  });

  it('renders unexpected error page when error is not 404', () => {
    const error = new Error('Test error message');
    (useRouteError as jest.Mock).mockReturnValue(error);
    (isRouteErrorResponse as unknown as jest.Mock).mockReturnValue(false);

    render(
      <MemoryRouter>
        <ErrorPage />
      </MemoryRouter>
    );

    expect(screen.getByText('Unexpected Application Error!')).toBeInTheDocument();
    expect(screen.getByText('Something went wrong, please try again.')).toBeInTheDocument();
    expect(screen.getByText('Test error message')).toBeInTheDocument(); // Error message
    expect(screen.getByText(/Error: Test error message/i)).toBeInTheDocument(); // Stack trace
    expect(screen.getByRole('link', { name: /take me back to safe ground/i })).toBeInTheDocument();
  });

  it('toggles error message visibility on button click in unexpected error page', () => {
    const error = new Error('Test error message');
    (useRouteError as jest.Mock).mockReturnValue(error);
    (isRouteErrorResponse as unknown as jest.Mock).mockReturnValue(false);

    // Render the ErrorPage inside MemoryRouter
    render(
      <MemoryRouter>
        <ErrorPage />
      </MemoryRouter>
    );

    // Ensure the error message is initially visible
    const toggleButton = screen.getByRole('button', { name: /hide error message/i });
    expect(screen.getByText('Test error message')).toBeVisible();

    // Click to hide the error message
    fireEvent.click(toggleButton);
    expect(screen.queryByText('Test error message')).not.toBeVisible();

    // Click to show the error message again
    fireEvent.click(screen.getByRole('button', { name: /open error message/i }));
    expect(screen.getByText('Test error message')).toBeVisible();
  });

  it('matches snapshot', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <ErrorPage />
      </MemoryRouter>
    );
    const attributesToRemove = document.body.querySelectorAll('div [id^="mantine"]');
    attributesToRemove.forEach((element) => {
      element.removeAttribute('id');
      element.removeAttribute('aria-describedby');
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
