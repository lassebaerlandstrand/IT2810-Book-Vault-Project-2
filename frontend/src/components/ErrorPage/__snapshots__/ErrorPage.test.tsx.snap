// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ErrorPage Component > matches snapshot 1`] = `
<DocumentFragment>
  <style
    data-mantine-styles="true"
  >
    :root{--mantine-font-family: Poppins, Roboto, Helvetica, Arial, sans-serif, Apple Color Emoji, Segoe UI Emoji;--mantine-font-family-headings: Poppins, Roboto, Helvetica, Arial, sans-serif, Apple Color Emoji, Segoe UI Emoji;--mantine-breakpoint-xs: 30em;--mantine-breakpoint-md: 64em;--mantine-breakpoint-xl: 90em;--mantine-breakpoint-xxs: 24em;--mantine-color-logo-theme-0: #fefefe;--mantine-color-logo-theme-1: #cbcfd2;--mantine-color-logo-theme-2: #97a0a6;--mantine-color-logo-theme-3: #7d8991;--mantine-color-logo-theme-4: #63717b;--mantine-color-logo-theme-5: #495a65;--mantine-color-logo-theme-6: #2f434f;--mantine-color-logo-theme-7: #132937;--mantine-color-logo-theme-8: #0b202e;--mantine-color-logo-theme-9: #061017;}:root[data-mantine-color-scheme="dark"]{--mantine-color-logo-theme-text: var(--mantine-color-logo-theme-4);--mantine-color-logo-theme-filled: var(--mantine-color-logo-theme-8);--mantine-color-logo-theme-filled-hover: var(--mantine-color-logo-theme-9);--mantine-color-logo-theme-light: rgba(47, 67, 79, 0.15);--mantine-color-logo-theme-light-hover: rgba(47, 67, 79, 0.2);--mantine-color-logo-theme-light-color: var(--mantine-color-logo-theme-3);--mantine-color-logo-theme-outline: var(--mantine-color-logo-theme-4);--mantine-color-logo-theme-outline-hover: rgba(99, 113, 123, 0.05);}:root[data-mantine-color-scheme="light"]{--mantine-color-logo-theme-text: var(--mantine-color-logo-theme-filled);--mantine-color-logo-theme-filled: var(--mantine-color-logo-theme-6);--mantine-color-logo-theme-filled-hover: var(--mantine-color-logo-theme-7);--mantine-color-logo-theme-light: rgba(47, 67, 79, 0.1);--mantine-color-logo-theme-light-hover: rgba(47, 67, 79, 0.12);--mantine-color-logo-theme-light-color: var(--mantine-color-logo-theme-6);--mantine-color-logo-theme-outline: var(--mantine-color-logo-theme-6);--mantine-color-logo-theme-outline-hover: rgba(47, 67, 79, 0.05);}
  </style>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 29.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 30em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 63.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 64em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 89.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 90em) {.mantine-hidden-from-xl {display: none !important;}}@media (max-width: 23.99375em) {.mantine-visible-from-xxs {display: none !important;}}@media (min-width: 24em) {.mantine-hidden-from-xxs {display: none !important;}}
  </style>
  <div
    class="_container_0b6ffd m_7485cace mantine-Container-root"
  >
    <div
      class="_content_0b6ffd m_7485cace mantine-Container-root"
    >
      <h1
        class="_title_0b6ffd m_8a5d1357 mantine-Title-root"
        data-order="1"
        style="--title-fw: var(--mantine-h1-font-weight); --title-lh: var(--mantine-h1-line-height); --title-fz: var(--mantine-h1-font-size);"
      >
        Unexpected Application Error!
      </h1>
      <p
        class="mantine-focus-auto _description_0b6ffd m_b6d8b162 mantine-Text-root"
        data-size="lg"
        style="--text-fz: var(--mantine-font-size-lg); --text-lh: var(--mantine-line-height-lg); text-align: center;"
      >
        Something went wrong, please try again.
      </p>
      <div
        class="m_4451eb3a mantine-Center-root"
      >
        <a
          href="/"
        >
          <button
            class="mantine-focus-auto mantine-active m_77c9d27d mantine-Button-root m_87cf2631 mantine-UnstyledButton-root"
            data-size="md"
            style="--button-height: var(--button-height-md); --button-padding-x: var(--button-padding-x-md); --button-fz: var(--mantine-font-size-md); --button-color: var(--mantine-color-white);"
            type="button"
          >
            <span
              class="m_80f1301b mantine-Button-inner"
            >
              <span
                class="m_811560b9 mantine-Button-label"
              >
                Take me back to safe ground
              </span>
            </span>
          </button>
        </a>
      </div>
      <div
        class="m_4451eb3a mantine-Center-root"
      >
        <div
          class="m_6d731127 mantine-Stack-root"
          style="--stack-gap: var(--mantine-spacing-md); --stack-align: center; --stack-justify: flex-start;"
        >
          <button
            class="mantine-focus-auto mantine-active m_77c9d27d mantine-Button-root m_87cf2631 mantine-UnstyledButton-root"
            style="--button-color: var(--mantine-color-white);"
            type="button"
          >
            <span
              class="m_80f1301b mantine-Button-inner"
            >
              <span
                class="m_811560b9 mantine-Button-label"
              >
                Hide error message
              </span>
            </span>
          </button>
          <div
            aria-hidden="false"
            class=""
            style="box-sizing: border-box; opacity: 1; transition: opacity 200ms ease;"
          >
            <pre
              class="m_b183c0a2 mantine-Code-root"
              data-block="true"
              dir="ltr"
            >
              Test error message
            </pre>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
